### ISSUE TITLE
gr_mpoly_divides_heap_threaded is not thread-safe on weakly-ordered CPUs (spurious GR_DOMAIN on Apple M1)

### ISSUE THREAD

**Issue by vneiger** [flintlib/flint#2849]
Issue prepared with Claude Opus after noticing a CI failure recently. (I don't feel familiar enough with this piece of FLINT to try to guide Claude towards a good solution.)

## Symptom

On Apple M1, `make check` fails in `gr_mpoly`:

```
FAIL: exact division incorrectly reported as GR_DOMAIN
i = 405, j = 0, dstatus1 = 0, dstatus2 = 1
Integer ring (fmpz)
```

Here the serial `gr_mpoly_divides_heap` returns `GR_SUCCESS` on `h = f*g` divided by `g`, while `gr_mpoly_divides_heap_threaded` returns `GR_DOMAIN`. The same test passes every time on x86-64.

## Cause

The threaded division is a port of `fmpz_mpoly_divides_heap_threaded`. That algorithm shares data between threads without locks and is only correct under x86's strong (TSO) memory ordering. That is why the `fmpz_mpoly` and `nmod_mpoly` versions are compiled only `#if FLINT_KNOW_STRONG_ORDER`. The `gr_mpoly` version has no such guard.

f2cdc22 ("Atomic fences for ARM thread safety") added a release/acquire pair around `polyQ->length`. That covers publishing new quotient terms, but not the **handoff of the producer role** between chunks (`src/gr_mpoly/divides_heap_threaded.c`, `trychunk`):

```c
/* thread A, finishing chunk L */
H->cur = next;
if (next != NULL)
    next->producer = 1;          /* plain store, no release */

/* thread B, concurrently inside trychunk(next) */
if (L->producer == 1)            /* plain load, no acquire */
{
    q_prev_length = Q->length;   /* may be stale on ARM */
    ...
```

On ARM, B can see `producer == 1` and still read a `Q->length` older than the one A published. B then computes its chunk's remainder from an incomplete quotient. `_gr_mpoly_divides_stripe` finds that remainder is not divisible and returns `GR_DOMAIN`. In principle the same race could also give a wrong quotient.

ThreadSanitizer (x86-64, fmpz, 4 threads) reports exactly these pairs as data races: `next->producer = 1` vs the `L->producer` reads, `H->cur = next` vs `L = H->cur` in `worker_loop`, and `H->length--` racing with itself.

`divrem_ideal_heap_threaded.c` (`ideal_trychunk`) has the same handoff. `divrem_heap_threaded.c` reuses `trychunk`/`worker_loop`, so it is affected too. Since `gr_mpoly_divides` dispatches to the threaded code, ordinary calls can hit this on ARM, not only the test.

## Possible fixes

1. **Add the missing fences (minimal).** Put `atomic_thread_fence(memory_order_acquire)` at the start of the `if (L->producer == 1)` branch, before `Q->length` is re-read. Put `atomic_thread_fence(memory_order_release)` before `next->producer = 1`. Move `H->length--` / `H->cur = next` under `H->mutex`. Do the same in `ideal_trychunk`. This is about 40 lines and keeps threading on ARM. A patch is available.
2. **Use real atomics (cleaner).** Declare the words shared without locks (`polyQ->length`, `L->producer`, `H->cur`, `H->failed`) as `_Atomic`, and use `atomic_load_explicit`/`atomic_store_explicit` with acquire/release. Plain accesses plus fences work in practice with gcc/clang, but they are still data races (UB) in C11 and keep showing up in TSan. This would also make the design reusable for `fmpz_mpoly`/`nmod_mpoly` (see the TODO in `gr_mpoly_ts_append`).
3. **Disable on weak ordering (conservative).** Wrap the `gr_mpoly` threaded division routines in `#if FLINT_KNOW_STRONG_ORDER` and fall back to the serial code otherwise, as `fmpz_mpoly`/`nmod_mpoly` already do. This is the quickest way to get CI green, but it gives up threaded mpoly division on Apple silicon and other ARM machines.

I'd suggest (1) now, or (3) if CI must be green immediately, with (2) as a follow-up. Any fix can only really be checked on ARM hardware, since the test never fails on x86.
